src/redux/sagas/expenditures/finalizeExpenditure.ts

```diff
@@ -95,12 +131,118 @@ function* finalizeExpenditureAction({
   }
   [finalizeExpenditure, annotateFinalizeExpenditure].forEach((channel) =>
     channel.channel.close(),
   );
+}
+
+type FinalizeExpenditureWithPermissions = {
+  colonyAddress: Address;
+  expenditure: Expenditure;
+  meta: Action<ActionTypes.EXPENDITURE_FINALIZE>['meta'];
+  annotationMessage?: string;
+  userAddress: Address;
+};
+
+function* finalizeExpenditureWithPermissions({
+  colonyAddress,
+  expenditure,
+  meta,
+  annotationMessage,
+  userAddress,
+}: FinalizeExpenditureWithPermissions) {
+  const batchKey = 'finalizeExpenditure';
+  const colonyManager: ColonyManager = yield getColonyManager();
+  const {
+    finalizeExpenditure,
+    annotateFinalizeExpenditure,
+  }: Record<string, ChannelDefinition> = yield createTransactionChannels(
+    meta.id,
+    ['finalizeExpenditure', 'annotateFinalizeExpenditure'],
+  );
 
+  try {
+    const colonyClient = yield colonyManager.getClient(
+      ClientType.ColonyClient,
+      colonyAddress,
+    );
+
+    const [permissionDomainId, childSkillIndex] = yield getPermissionProofs(
+      colonyClient.networkClient,
+      colonyClient,
+      expenditure.nativeDomainId,
+      ColonyRole.Arbitration,
+      userAddress,
+    );
+
+    const params = [permissionDomainId, childSkillIndex, expenditure.nativeId];
+
+    yield fork(createTransaction, finalizeExpenditure.id, {
+      context: ClientType.ColonyClient,
+      methodName: 'finalizeExpenditureViaArbitration',
+      identifier: colonyAddress,
+      params,
+      group: {
+        key: batchKey,
+        id: meta.id,
+        index: 0,
+      },
+    });
+
+    if (annotationMessage) {
+      yield fork(createTransaction, annotateFinalizeExpenditure.id, {
+        context: ClientType.ColonyClient,
+        methodName: 'annotateTransaction',
+        identifier: colonyAddress,
+        group: {
+          key: batchKey,
+          id: meta.id,
+          index: 1,
+        },
+        ready: false,
+      });
+    }
+
+    yield takeFrom(
+      finalizeExpenditure.channel,
+      ActionTypes.TRANSACTION_CREATED,
+    );
+    if (annotationMessage) {
+      yield takeFrom(
+        annotateFinalizeExpenditure.channel,
+        ActionTypes.TRANSACTION_CREATED,
+      );
+    }
+
+    yield initiateTransaction({ id: finalizeExpenditure.id });
+
+    const {
+      payload: {
+        receipt: { transactionHash: txHash },
+      },
+    } = yield waitForTxResult(finalizeExpenditure.channel);
+
+    if (annotationMessage) {
+      yield uploadAnnotation({
+        txChannel: annotateFinalizeExpenditure,
+        message: annotationMessage,
+        txHash,
+      });
+    }
+
+    yield put<AllActions>({
+      type: ActionTypes.EXPENDITURE_FINALIZE_SUCCESS,
+      payload: {},
+      meta,
+    });
+  } catch (error) {
+    return yield putError(ActionTypes.EXPENDITURE_FINALIZE_ERROR, error, meta);
+  }
+  [finalizeExpenditure, annotateFinalizeExpenditure].forEach((channel) =>
+    channel.channel.close(),
+  );
```

Comment:
It feels like we've ended up with way more code (and duplications) by splitting it into helpers (all because of annotations) - maybe having an if inside the main saga instead would make it a bit more readable?